lib/types.js

```diff
@@ -2224,7 +2224,7 @@ RecordType.prototype._createChecker = function () {
       body += '    p[j] = \'' + field.name + '\';\n';
       body += '    b &= ';
       if (field.defaultValue() === undefined) {
-        body += 't' + i + '._check(v.' + field.name + ', f, h, p);\n';
+        body += 't' + i + '._check((v.' + field.name + (field.aliases != null && field.aliases.length > 0 ? ' || v.' + (Array.isArray(field.aliases) ?  field.aliases.join(' || v.') : field.aliases) : '') + '), f, h, p);\n';
```

Comment:
Thanks for getting back to me! I was unaware of any type casting between aliases according to the official spec. Did you have a structure/format/idea for how to separate it and/or where that code should be?

Reply:
It would be nice if we could mimic the resolver API for binary encodings, ending with something like:

```js
var writerType = Type.forSchema({ // Original type.
  name: 'Thing',
  type: 'record',
  fields: [
    {name: 'before', type: 'int'},
  ],
});

var readerType = Type.forSchema({ // Type we are reading as.
  name: 'Thing',
  type: 'record',
  fields: [
    {name: 'after', aliases: ['before'], type: 'long'},
  ],
});

var resolver = readerType.createResolver(writerType); // Same as now.
var decoded = readerType.fromString('{"before": 123}', resolver); // {after: 123}
```

We can actually delegate the heavy lifting to the existing (binary encoding) resolution logic. Binary de/encoding is very fast so the implementation would just need to:

1. Decode the JSON-encoded data using the writer type;
2. Binary-encode the result of 1 using the writer type;
3. Use the existing resolver functionality to decode the result of 2.

This could probably fit within the `fromString` method; we would just need to add the writer type to the resolver (IIRC right now it only has the reader's). Let me know if this is unclear.

Note that `avsc` currently reads and writes JSON-encoding directly as strings (`fromString` in the example above reads a JSON string). If you think it would be worth it, we could add convenience methods that deal with objects, saving `JSON.parse` and `JSON.stringify` calls.

Reply:
I do think that saving parse/stringify calls would be helpful. The issue with doing it via binary encoding is that sometimes the writer schema is unknown, as we are sent plain json(could even be free form, not validated just generated according to spec). That is why being able to assume the alias convention from only the latest reader schema would be helpful. I see the issue with binary encoding, however JSON could side step the issue of offsets of values being different, because we have the keys.

Reply:
Agreed that it would be useful to have a more flexible way to read JSON data. These requirements don't match the spec's [definition of aliases](http://avro.apache.org/docs/current/spec.html#Aliases) however, so I don't think it makes as much sense to implement here.

What do you think about implementing this as a separate package? You would be free to design an API that best suits your use-case without being tied to the spec. You could still use the `aliases` field, overloading its meaning. I'm happy to help review the code too.

In parallel, I'll try to find some time to implement standard alias resolution logic for JSON, as well as convenience `toJSON` and `fromJSON` methods.